addons/hw_drivers/controllers/driver.py

```diff
@@ -358,6 +358,8 @@ def run(self):
             devices = updated_devices
             for path in [device_rm for device_rm in removed if device_rm in iot_devices]:
                 iot_devices[path].disconnect()
+                self.send_alldevices()
+                _logger.info('For device %s will be remove', path)
```

Comment:
i would call send_alldevices out of the loop, because if you have two removed devices at this time you'll all it twice and it'll make redundant operations at the other end

``` python
            disconnected_devices = False
            for path in [device_rm for device_rm in removed if device_rm in iot_devices]:
                iot_devices[path].disconnect()
                disconnected_devices = True
            # If any devices was disconnected, resend all the connected devices to the server.
            if disconnected_devices:
                self.send_alldevices()
```

Reply:
better way: It resolve similar issue with added device.

`if added or removed:
    self.send_alldevices()`

Reply:
true it's better, you could do it just before time.sleep(3) 👍

Reply:
and remove send_alldevices line 371

Reply:
it's already done :)